worker/changestream/doc.go

```diff
@@ -0,0 +1,106 @@
+//	│ UUID A  │ │ UUID B  │ │ UUID A  │ │ UUID C  │
+//	│         │ │         │ │         │ │         │
+//	└─────────┘ └─────────┘ └─────────┘ └─────────┘
+//
+//	┌─────────┐ ┌─────────┐ ┌─────────┐
+//	│         │ │         │ │         │
+//	│ NS   1  │ │ NS   1  │ │ NS   1  │
+//	│ UUID B  │ │ UUID A  │ │ UUID C  │
+//	│         │ │         │ │         │
+//	└─────────┘ └─────────┘ └─────────┘
+//
+// No additional information about the change is provided in the event, other
+// than:
+//
+//   - The change type (CREATE, UPDATE, DELETE)
+//   - The namespace (generally the table name, but not guaranteed)
+//   - The changed_uuid (the primary key of the row that changed)
+//
+// It's not useful to provide additional information about the change, because
+// in an eventual system, the change may be considered stale by the time it
+// reaches the subscriber. The subscriber will need to query the database to
+// retrieve the latest information about the change. They may already have
+// witnessed the change so may be ignored at that point.
+//
+// The event multiplexer will send the changes to the appropriate subscription.
+// This involves grouping changes into terms. Terms are a set of changes that
+// are sent to the subscriber in a single batch. Terms are used to ensure that
+// that there is a consistent view of the database at the subscriber.
+// The event multiplexer will wait for all terms to be read by subscribers
+// before attempting to send the next term. All terms will be sent in a
+// async manner, allowing all subscribers to worker on the same term at the same
+// time. Once a term is marked as done, then and only then is another term
+// expected to be worked on.
+//
+// Terms are expected to be consumed as fast as possible. If a term can not be
+// processed or is taking to long to consume by a subscriber then the subscriber
+// will be unsubscribed from the next term. This is to ensure that the event
+// multiplexer does not get blocked by a slow subscriber.
+//
+//	┌───────────┐ ┌───────────┐ ┌───────────────────────────┐ ┌───────────┐
+//	│           │ │           │ │                           │ │           │
+//	│  TERM 1   │ │  TERM 2   │ │  TERM 3                   │ │  TERM 4   │
+//	│           │ │           │ │                           │ │           │
+//	└───────────┘ └───────────┘ └───────────────────────────┘ └───────────┘
+//
+// In the above example, the event multiplexer will wait for all subscribers to
+// consume TERM 1 before sending TERM 2. If a subscriber is slow to consume TERM
+// 3, then the subscriber will be unsubscribed from consecutive terms. The
+// subscriber will be notified of the unsubscribe event. They can then decide
+// to either resubscribe or start from a fresh state and try and keep up.
+// This model will ensure that one subscriber doesn't attempt to bring down the
+// entire system.
+//
+// If a subscriber can't consume or process a term fast enough, it is then
```

Comment:
I assume we'll be adding lots-o-metrics (tm) to record this type of occurrence

Reply:
You are indeed correct.